Installation docs (#3)
openforcefield/openff-docs: source/install.md

If you don't have Conda or Mamba installed, installing MambaForge will give you access to everything you need to install OpenFF software. If you can, we recommend installing MambaForge locally to your user account, rather than system-wide, so that you can freely create and destroy environments and manage your own configuration. If something goes wrong, you can always delete your MambaForge installation and start again, and you'll only lose your installed software.

:::{hint}
If you're running Mamba through MambaForge, you're configured to use the Conda Forge channel by default, and you don't need to pass the `-c conda-forge` argument to the commands in this page. Including this argument does no harm, so we've included it for the benefit of users that don't have this configuration.
:::

[Mamba]: https://mamba.readthedocs.io/en/latest/index.html
[MambaForge]: https://github.com/conda-forge/miniforge#mambaforge
[MambaForge repository]: https://github.com/conda-forge/miniforge#mambaforge
[Mambaforge-MacOSX-x86_64]: https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-MacOSX-x86_64.sh

> [Mambaforge-Linux-x86_64]: https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh

Author: Good call, I'll figure out a way to do that. I figured if they're happy to share curl commands with name interpolation of these URLs they can't be too likely to change!

> [Mambaforge-Linux-x86_64]: https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh

Author: Added a CI workflow that will tell us if any links break in the built documentation!

(conda_init)=
### To init or not to init

After it's done installing, the MambaForge installer will ask you if you want it to initialize MambaForge. Initialization involves adding a section to your shell startup script that tells the shell where to find Conda/Mamba. Answering "no" will leave you startup scripts untouched, but you will need to run an additional command every time you want to use Conda/Mamba or a Conda environment. Answering "yes" will attempt to add a version of this command to your shell startup script, which generally makes Conda easier to use.

If you do answer "yes", we recommend preventing Conda from activating the base environment in every new shell:

```shell-session
$ conda config --set auto_activate_base false
```

> 

Author: I've been burned a couple of times by systems failing to update because a package manager was expecting the system Python and got the base Conda environment instead. I know most package managers are more resilient than that but I really don't want to create weird problems for our users.

(existing_conda)=
### Installing Mamba in an existing Conda installation

If you already have Conda installed, you don't need to install MambaForge; you can install Mamba on its own with:

```shell
conda install -c conda-forge mamba
```

Or you can simply replace any calls to `mamba` with an identical call to `conda`:

```diff
- mamba env create -c conda-forge -n openff -f environment.yaml
+ conda env create -c conda-forge -n openff -f environment.yaml
```

We recommend installing Mamba, as it can solve OpenFF environments in seconds where Conda would take minutes or hours.

(managing_environments)=
## Managing Environments

Conda and Mamba install software into [virtual environments]. These environments include installations of some collection of software and all their dependencies in an isolated group that doesn't affect the software installed in the rest of your computer system. This means you can install software that's incompatible with what's already installed; you just put the new software in its own environment!

> 

Reviewer: Suggest adding something like

> Virtual environments are designed to be rapidly deployable, which also means they can quickly and safely be removed and re-created in a matter of seconds in case something goes wrong.

Personal preference to emphasize some smithed version of this, I regularly find that beginners waste days of their lives trying to fix weird states when blowing everything away and starting fresh takes a few minutes.

Reviewer: Oh, maybe the following paragraph is a better location ...

Author: Done in 86c1c84

```

Or from an [environment file], which can specify both channels and dependencies:

```shell-session
$ mamba env create -n <name> -f <path to .yaml file>
```

Any time you want to use software installed in an environment you have created, you must activate it. An environment will stay active until you close the terminal window or shell session:

```shell-session
$ conda activate <name>
$ jupyter notebook .
```

[virtual environments]: conda:user-guide/concepts/environments
[local to your user]: install_mamba

### Advanced Environment Management

If you prefer, you can specify the path to a virtual environment's root directory as a prefix instead of giving an environment's name:

```shell-session
$ mamba create -c conda-forge -p <prefix> <package(s)>
$ conda activate <prefix>
```

If you specify neither the name `-n` nor the prefix `-p`, the current active environment will be used.

New software can be installed in an environment by listing packages:

```shell-session
$ mamba install -n <name> -c conda-forge <package(s)>
```

Or by adding the dependencies from an environment file:

```shell-session
$ mamba env update -n <name> -f <path to .yaml file>
```

> Mamba/Conda cache the packages you install, so if you've already downloaded a particular version, you can install it in as many environments as you like without having to re-download it. In fact, Mamba/Conda try their hardest to unpack each version once, and then hardlink to it from each environment to keep the storage size of environments down. Be aware of this if you like to tinker with the code of your dependencies - your changes may affect other environments!

Reviewer: > tinker with ...

If this implies "don't muck around in `/site-packages/`" we should have stronger language against doing this. If you're referring to editable installs while switching branches between projects ... ugh, that's annoying and I keep doing it.

Author: Yeah referring to the former, I'll strengthen and clarify.

Author: Done in 31bd01f

The software in an environment can be automatically upgraded:

```shell-session
$ mamba upgrade -n <name> --all
```

> But that can go wrong if you're mixing channels. Recreating the environment often works better:

Reviewer: ```suggestion
But that can go wrong if you're mixing channels and might be slow if the environment is old. Recreating the environment often works better:
```

```shell-session
$ mamba env remove -n <name>
$ mamba create -n <name> -c conda-forge <package(s)>
```

Or specifying the particular package you want to update:

```shell-session
$ mamba upgrade -n <name> <package>
```

### Combining channels

Sometimes, some of the software you want to use is not available on Conda Forge. In these cases, you can either install it in the usual way for your OS, or you can try combining channels:

```shell-session
$ mamba create -n psi4_and_bespokefit -c psi4 -c conda-forge -c anaconda openff-bespokefit psi4
```

> In this case, this doesn't work because Conda/Mamba has to be very clever about how it solves this environment. It has to be able to choose versions of dependencies from the `anaconda` channel, even when those dependencies exist in the higher-priority `psi4` channel. It's generally recommended to use [strict channel priority] to avoid conflicts, but sometimes an environment can only be solved when channel priority is set to `"flexible"`. The catch is that the environment may solve and install but not work! But it's usually worth a try.

Reviewer: > The catch is that the environment may solve and install but not work!

Is this true? I'm not sure I've ever had this happen to me. Usually my issues with mixed channels relate to the environment not solving at all.

Author: I'm sure I had a problem last week where a BespokeFit+Psi4 environment solved with flexible priority, but then I got missing symbol errors when trying to run it. And I believe this is one of the motivations behind strict. But it's working today so maybe I'm misremembering.

Reviewer 2: Hi There. FYI I had a day off yesterday, so only just starting to look at this. I will send my comments to Josh by email.

This environment is also complex enough that remembering all the channels every time you want to update it is risky. So instead of trying to create it in one line, we might be better off creating it empty, configuring its channel list and priorities, and then installing our software:

```shell-session
$ mamba create -n psi4_and_bespokefit
$ conda activate psi4_and_bespokefit
$ conda config --env --set channel_priority flexible
$ conda config --env --prepend channels anaconda
$ conda config --env --prepend channels conda-forge
$ conda config --env --prepend channels psi4
$ mamba install openff-bespokefit psi4 ambertools
```

[environment file]: conda:create-env-file-manually
[strict channel priority]: conda:concepts-performance-channel-priority

(install_windows)=
## OpenFF on Windows

> OpenFF does not directly support Windows operating systems, though much of our software is pure Python code and may work anyway as long as you can provide our dependencies.

Reviewer: ```suggestion
OpenFF does not support or test on Windows natively. All of our software is pure Python code and may work anyway as long as you can provide our dependencies, but many of these dependencies are not available on Windows and may not be in the foreseeable future.
```

Reviewer: Personally prefer emphasizing that we CURRENTLY cannot do anything on Windows because of our upstreams, even though this is a cop-out it shifts the blame away from us and puts a bit of distance between us and the appearance that Windows is not supported because we didn't bother trying.

Author: Good call - I love how your phrasing warns that users are gonna have a bad time without feeling like we're saying they're breaking an arbitrary rule.

> Regardless, we recommend using the [Windows Subsystem for Linux] to run OpenFF software on Windows. WSL runs a Linux kernel within your Windows system so you can run ordinary software as if you had a Linux system. If your hardware supports it, we suggest using WSL2 for a smoother experience; this is the default for new WSL installations on supported hardware. WSL2 requires hardware virtualization support, which is available on most modern CPUs but may require activation in the BIOS/UEFI. 

Reviewer: ```suggestion
Instead, we recommend using the [Windows Subsystem for Linux] to run OpenFF software on Windows. WSL runs a Linux kernel within your Windows system so you can run ordinary software as if you had a Linux system. If your hardware supports it, we suggest using WSL2 for a smoother experience; this is the default for new WSL installations on supported hardware. WSL2 requires hardware virtualization support, which is available on most modern CPUs but may require activation in the BIOS/UEFI. 
```
